File: apps/api/app/api/routes/requirement_analysis.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, ConfigDict

from app.db.session import get_session
from app.requirement_analysis.models import RequirementAnalysisSessionCreate, RequirementAnalysisTurnCreate
from app.requirement_analysis.session_application_service import RequirementAnalysisApplicationService


router = APIRouter(prefix="/requirement-analysis", tags=["requirement-analysis"])


class RequirementAnalysisTemplateSave(BaseModel):
    model_config = ConfigDict(extra="forbid")

    content: str
    name: str | None = None
    description: str | None = None


class RequirementAnalysisTemplateCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    base_template_id: str
    name: str
    description: str = ""


def get_requirement_analysis_service(session=Depends(get_session)) -> RequirementAnalysisApplicationService:
    return RequirementAnalysisApplicationService(session)


def _bad_request(exc: ValueError) -> HTTPException:
    return HTTPException(status_code=400, detail=str(exc))


def _not_found(message: str) -> HTTPException:
    return HTTPException(status_code=404, detail=message)


@router.get("/lab-config")
def get_requirement_analysis_lab_config(
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    return service.get_lab_config()


@router.get("/orchestrators")
def list_requirement_analysis_orchestrators(
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    return service.list_orchestrators()


@router.post("/orchestrators/reload")
def reload_requirement_analysis_orchestrators(
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    try:
        return service.reload_orchestrators()
    except ValueError as exc:
        raise _bad_request(exc) from exc


@router.get("/providers")
def list_requirement_analysis_providers(
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    return service.list_providers()


@router.get("/templates")
def list_requirement_analysis_templates(
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    return service.list_templates()


@router.get("/template-bases")
def list_requirement_analysis_template_bases(
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    return service.list_base_templates()


@router.post("/templates")
def create_requirement_analysis_template(
    payload: RequirementAnalysisTemplateCreate,
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    template = service.create_template(
        base_template_id=payload.base_template_id,
        name=payload.name,
        description=payload.description,
    )
    if template is None:
        raise _not_found("Requirement Analysis base template not found")
    return template


@router.get("/templates/{template_id}")
def get_requirement_analysis_template(
    template_id: str,
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    template = service.get_template(template_id)
    if template is None:
        raise _not_found("Requirement Analysis template not found")
    return template


@router.put("/templates/{template_id}")
def save_requirement_analysis_template(
    template_id: str,
    payload: RequirementAnalysisTemplateSave,
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    template = service.save_template(
        template_id,
        payload.content,
        name=payload.name,
        description=payload.description,
    )
    if template is None:
        raise _not_found("Requirement Analysis template not found")
    return template


@router.delete("/templates/{template_id}")
def delete_requirement_analysis_template(
    template_id: str,
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    result = service.delete_template(template_id)
    if result is None:
        raise _not_found("Requirement Analysis template not found")
    return result


@router.post("/sessions")
def create_requirement_analysis_session(
    payload: RequirementAnalysisSessionCreate,
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    try:
        return service.create_session(payload)
    except ValueError as exc:
        raise _bad_request(exc) from exc


@router.get("/sessions/{session_id}")
def get_requirement_analysis_session(
    session_id: str,
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    session = service.get_session(session_id)
    if session is None:
        raise _not_found("Requirement Analysis session not found")
    return session


@router.post("/sessions/{session_id}/turns")
def create_requirement_analysis_turn(
    session_id: str,
    payload: RequirementAnalysisTurnCreate,
    service: RequirementAnalysisApplicationService = Depends(get_requirement_analysis_service),
):
    turn = service.add_turn(session_id, payload)
    if turn is None:
        raise _not_found("Requirement Analysis session not found")
    return turn

File: apps/web/src/lib/requirementAnalysis.ts
import { api } from "./api";
import type {
  RequirementAnalysisLabConfig,
  RequirementAnalysisOrchestratorEnvelope,
  RequirementAnalysisProviderEnvelope,
  RequirementAnalysisSession,
  RequirementAnalysisSessionCreateInput,
  RequirementAnalysisTemplateDetail,
  RequirementAnalysisTemplateEnvelope,
  RequirementAnalysisTurnEnvelope,
  RequirementSpecWorkItem,
  RequirementSpecWorkItemConfigureInput,
  RequirementSpecWorkItemCreateInput,
  RequirementSpecWorkItemEnvelope,
  RequirementSpecWorkItemSaveAsInput,
  RequirementSpecWorkItemSaveSessionArtifactsInput,
} from "./api";

export function getRequirementAnalysisLabConfig() {
  return api.get<RequirementAnalysisLabConfig>("/requirement-analysis/lab-config");
}

export function getRequirementAnalysisOrchestrators() {
  return api.get<RequirementAnalysisOrchestratorEnvelope>("/requirement-analysis/orchestrators");
}

export function reloadRequirementAnalysisOrchestrators() {
  return api.post<RequirementAnalysisOrchestratorEnvelope>("/requirement-analysis/orchestrators/reload");
}

export function getRequirementAnalysisProviders() {
  return api.get<RequirementAnalysisProviderEnvelope>("/requirement-analysis/providers");
}

export function getRequirementAnalysisTemplates() {
  return api.get<RequirementAnalysisTemplateEnvelope>("/requirement-analysis/templates");
}

export function getRequirementAnalysisTemplateBases() {
  return api.get<RequirementAnalysisTemplateEnvelope>("/requirement-analysis/template-bases");
}

export function getRequirementSpecWorkItems() {
  return api.get<RequirementSpecWorkItemEnvelope>("/requirement-analysis/spec-items");
}

export function createRequirementSpecWorkItem(payload: RequirementSpecWorkItemCreateInput) {
  return api.post<RequirementSpecWorkItem>("/requirement-analysis/spec-items", payload);
}

export function configureRequirementSpecWorkItem(specItemId: string, payload: RequirementSpecWorkItemConfigureInput) {
  return api.post<RequirementSpecWorkItem>(`/requirement-analysis/spec-items/${specItemId}/configure`, payload);
}

export function publishRequirementSpecWorkItem(specItemId: string) {
  return api.post<RequirementSpecWorkItem>(`/requirement-analysis/spec-items/${specItemId}/publish`);
}

export function deleteRequirementSpecWorkItem(specItemId: string) {
  return api.delete<{ deleted: boolean; spec_item_id: string }>(`/requirement-analysis/spec-items/${specItemId}`);
}

export function saveRequirementSpecWorkItemSessionArtifacts(specItemId: string, payload?: RequirementSpecWorkItemSaveSessionArtifactsInput) {
  return api.post<RequirementSpecWorkItem>(`/requirement-analysis/spec-items/${specItemId}/save-session-artifacts`, payload ?? {});
}

export function saveRequirementSpecWorkItemSessionArtifactsAs(specItemId: string, payload: RequirementSpecWorkItemSaveAsInput) {
  return api.post<RequirementSpecWorkItem>(`/requirement-analysis/spec-items/${specItemId}/save-session-artifacts-as`, payload);
}

export function getRequirementAnalysisTemplate(templateId: string) {
  return api.get<RequirementAnalysisTemplateDetail>(`/requirement-analysis/templates/${templateId}`);
}

export function createRequirementAnalysisTemplate(baseTemplateId: string, name: string, description: string) {
  return api.post<RequirementAnalysisTemplateDetail>("/requirement-analysis/templates", {
    base_template_id: baseTemplateId,
    name,
    description,
  });
}

export function saveRequirementAnalysisTemplate(templateId: string, content: string, name?: string, description?: string) {
  return api.put<RequirementAnalysisTemplateDetail>(`/requirement-analysis/templates/${templateId}`, {
    content,
    name,
    description,
  });
}

export function deleteRequirementAnalysisTemplate(templateId: string) {
  return api.delete<{ deleted: boolean; template_id: string }>(`/requirement-analysis/templates/${templateId}`);
}

export function saveRequirementAnalysisTemplateAsBase(templateId: string) {
  return api.post<RequirementAnalysisTemplateDetail>(`/requirement-analysis/templates/${templateId}/save-as-base`);
}

export function createRequirementAnalysisSession(payload: RequirementAnalysisSessionCreateInput) {
  return api.post<RequirementAnalysisSession>("/requirement-analysis/sessions", payload);
}

export function getRequirementAnalysisSession(sessionId: string) {
  return api.get<RequirementAnalysisSession>(`/requirement-analysis/sessions/${sessionId}`);
}

export function createRequirementAnalysisTurn(sessionId: string, userInput: string) {
  return api.post<RequirementAnalysisTurnEnvelope>(`/requirement-analysis/sessions/${sessionId}/turns`, {
    user_input: userInput,
  });
}
